src/libcrun/linux.c

```diff
@@ -2736,29 +2736,116 @@ libcrun_set_oom (libcrun_container_t *container, libcrun_error_t *err)
   return 0;
 }
 
+const char *sysctlRequiringIPC[] = {
+  "kernel/msgmax",
+  "kernel/msgmnb",
+  "kernel/msgmni",
+  "kernel/sem",
+  "kernel/shmall",
+  "kernel/shmmax",
+  "kernel/shmmni",
+  "kernel/shm_rmid_forced",
+  NULL
+};
+
+static int
+validate_sysctl (const char *original_value, const char *name, unsigned long namespaces_created, libcrun_error_t *err)
+{
+  const char *namespace = "";
+
+  if (has_prefix (name, "fs/queue/"))
+    {
+      if (namespaces_created & CLONE_NEWIPC)
+        return 0;
+
+      namespace = "IPC";
+      goto fail;
+    }
+
+  if (has_prefix (name, "kernel/"))
+    {
+      size_t i;
+
+      for (i = 0; sysctlRequiringIPC[i]; i++)
+        if (strcmp (sysctlRequiringIPC[i], name) == 0)
+          {
+            if (namespaces_created & CLONE_NEWIPC)
+              return 0;
+
+            namespace = "IPC";
+            goto fail;
+          }
+
+      if (strcmp (name, "kernel.domainname") == 0)
+        {
+          if (namespaces_created & CLONE_NEWUTS)
+            return 0;
+
+          namespace = "UTS";
+          goto fail;
+        }
+
+      if (strcmp (name, "kernel.hostname") == 0)
+        return crun_make_error (err, 0, "the sysctl `%s` conflicts with OCI field `hostname`", original_value);
+    }
+  if (has_prefix (name, "net/"))
+    {
+      if (namespaces_created & CLONE_NEWNET)
+        return 0;
```

Comment:
In runc, there's an additional check (for netns only, for some reason) that in case ns path is provided, we're not using host ns. This check prevents something like 
```
podman run --network=ns:/proc/self/ns/net --sysctl ...
```
from changing host net sysctls.

Perhaps you have this check somewhere else? Or do you think that this corner case is not worth checking for?

Reply:
we are not currently checking for that.

I was wondering why runc does it only for the network namespace, not for the others so I've not added (and honestly it seems a bit overkilling to me).

Reply:
I agree. Perhaps someone screwed up at some point and decided to add this. 

This was initially implemented by https://github.com/opencontainers/runc/pull/1138 in order to fix https://github.com/moby/moby/issues/27484 (i.e. prevent `docker run --net=host --sysctl net.xxx` from working).

I see that for Podman this is prevented by Podman:

[kir@kir-rhat runc]$ podman run --rm --runtime=runc --net=host --sysctl net.ipv4.ip_forward=hey alpine true
Error: sysctl net.ipv4.ip_forward=hey can't be set since Host Namespace set to host: invalid argument

(recently introduced by and I guess the same thing should be done by Docker (left a comment in there, https://github.com/moby/moby/issues/27484#issuecomment-1019017549)

Reply:
So yes, I'm OK with not checking this in crun.